polkadot/node/core/pvf/common/src/worker/security.rs

```diff
@@ -17,9 +17,129 @@
 //! Functionality for securing workers.
 //!
 //! This is needed because workers are used to compile and execute untrusted code (PVFs).
+//!
+//! We currently employ the following security measures:
+//!
+//! - Restrict filesystem
+//!   - Use Landlock to remove all unnecessary FS access rights.
+//!   - Unshare the user and mount namespaces.
+//!   - Change the root directory to a worker-specific temporary directory.
+//! - Remove env vars
+
+use crate::LOG_TARGET;
+
+/// Unshare the user namespace and change root to be the artifact directory.
+#[cfg(target_os = "linux")]
+pub fn unshare_user_namespace_and_change_root(
+	worker_dir_path: &std::path::Path,
+) -> Result<(), &'static str> {
+	use std::{ffi::CString, os::unix::ffi::OsStrExt, ptr};
+
+	let worker_dir_path_c = CString::new(worker_dir_path.as_os_str().as_bytes()).unwrap();
+	let root_c = CString::new("/").unwrap();
+	let dot_c = CString::new(".").unwrap();
+
+	// SAFETY: We pass null-terminated C strings and use the APIs as documented. In fact, steps (2)
+	//         and (3) are adapted from the example in pivot_root(2), with the additional change
+	//         described in the `pivot_root(".", ".")` section.
+	unsafe {
+		// 1. `unshare` the user and the mount namespaces.
+		//
+		// Separate calls: in case one flag succeeds but other fails, we give a more precise error.
+		if libc::unshare(libc::CLONE_NEWUSER) < 0 {
+			return Err("unshare user namespace")
+		}
+		if libc::unshare(libc::CLONE_NEWNS) < 0 {
+			return Err("unshare mount namespace")
+		}
```

Comment:
I didn't do it because on some systems one of these may fail (e.g. unsharing user namespaces is disabled), and I wanted an accurate error in that case. Not sure if it's a valid concern.

Reply:
The question is, do the separate error messages provide any meaningful info that cannot be provided in a single message? For us, I believe they don't. It doesn't matter which of the two calls failed, we cannot proceed anyway. Does that separation clear up anything for the user? You know best. If you can direct users more precisely to take some actions in each case, keep them separate (and probably make every message more specific with a link to external documentation on how to enable things). If it's just a "you should enable the whole thing" message, it's better to join them.

Reply:
> If you can direct users more precisely to take some actions in each case, keep them separate (and probably make every message more specific with a link to external documentation on how to enable things)

If they disabled it, they probably know how to enable it, or they can Google (or hopefully Kagi) for up-to-date info for their system. Embedding a link runs the risk of the link breaking after some time.

I would keep as-is. 🤷‍♂️

Reply:
Okay, but then, if the only possible reason for failure is that the user has consciously disabled the feature, it's okay to report in a single message?

(Or just leave it, we've already spent many more lines of text discussing that than the number of lines of code that could be saved here)

Reply:
Okay, you win. :P